In [1]:
from pulp import *
import pandas as pd

In [2]:
import os
os.chdir('C:\\Users\\varun\\Desktop\\GTA\\ISYE6501\\week 12 - optimization pulp Mcnemar')

### instantiate a problem class, name it “linearDiet” and we’re looking for an optimal minium cost so we use LpMinimize

In [3]:
prob = LpProblem("linearDiet", LpMinimize)

In [4]:
diet = pd.read_excel("diet.xls")
diet.head()


Foods  Price/ Serving      Serving Size  Calories  \
0      Frozen Broccoli            0.16         10 Oz Pkg      73.8   
1          Carrots,Raw            0.07  1/2 Cup Shredded      23.7   
2          Celery, Raw            0.04           1 Stalk       6.4   
3          Frozen Corn            0.18           1/2 Cup      72.2   
4  Lettuce,Iceberg,Raw            0.02            1 Leaf       2.6   

   Cholesterol mg  Total_Fat g  Sodium mg  Carbohydrates g  Dietary_Fiber g  \
0             0.0          0.8       68.2             13.6              8.5   
1             0.0          0.1       19.2              5.6              1.6   
2             0.0          0.1       34.8              1.5              0.7   
3             0.0          0.6        2.5             17.1              2.0   
4             0.0          0.0        1.8              0.4              0.3   

   Protein g  Vit_A IU  Vit_C IU  Calcium mg  Iron mg  
0        8.0    5867.4     160.2       159.0      2.3  
1        0.6   15471.0       5.1        14.9      0.3  
2        0.3      53.6       2.8        16.0      0.2  
3        2.5     106.6       5.2         3.3      0.3  
4        0.2      66.0       0.8         3.8      0.1

In [5]:
# Create a list of food items
food_items = list(diet['Foods'])

# Creating dictinary for each of the supplements against food item
costs = dict(zip(food_items,diet['Price/ Serving']))

calories = dict(zip(food_items,diet['Calories']))

fat = dict(zip(food_items,diet['Total_Fat g']))

cholesterol = dict(zip(food_items,diet['Cholesterol mg']))

carbs = dict(zip(food_items,diet['Carbohydrates g']))

fiber = dict(zip(food_items,diet['Dietary_Fiber g']))

sodium = dict(zip(food_items,diet['Sodium mg']))

protein = dict(zip(food_items,diet['Protein g']))

vit_A = dict(zip(food_items,diet['Vit_A IU']))

vit_C = dict(zip(food_items,diet['Vit_C IU']))

calcium = dict(zip(food_items,diet['Calcium mg']))

iron = dict(zip(food_items,diet['Iron mg']))

food_items

['Frozen Broccoli',
 'Carrots,Raw',
 'Celery, Raw',
 'Frozen Corn',
 'Lettuce,Iceberg,Raw',
 'Peppers, Sweet, Raw',
 'Potatoes, Baked',
 'Tofu',
 'Roasted Chicken',
 'Spaghetti W/ Sauce',
 'Tomato,Red,Ripe,Raw',
 'Apple,Raw,W/Skin',
 'Banana',
 'Grapes',
 'Kiwifruit,Raw,Fresh',
 'Oranges',
 'Bagels',
 'Wheat Bread',
 'White Bread',
 'Oatmeal Cookies',
 'Apple Pie',
 'Chocolate Chip Cookies',
 'Butter,Regular',
 'Cheddar Cheese',
 '3.3% Fat,Whole Milk',
 '2% Lowfat Milk',
 'Skim Milk',
 'Poached Eggs',
 'Scrambled Eggs',
 'Bologna,Turkey',
 'Frankfurter, Beef',
 'Ham,Sliced,Extralean',
 'Kielbasa,Prk',
 "Cap'N Crunch",
 'Cheerios',
 "Corn Flks, Kellogg'S",
 "Raisin Brn, Kellg'S",
 'Rice Krispies',
 'Special K',
 'Oatmeal',
 'Malt-O-Meal,Choc',
 'Pizza W/Pepperoni',
 'Taco',
 'Hamburger W/Toppings',
 'Hotdog, Plain',
 'Couscous',
 'White Rice',
 'Macaroni,Ckd',
 'Peanut Butter',
 'Pork',
 'Sardines in Oil',
 'White Tuna in Water',
 'Popcorn,Air-Popped',
 'Potato Chips,Bbqflvr',
 'Pretzel

### Declare decision variable using LpVariable class

In [6]:
# create varaibles
food_vars = LpVariable.dicts('food', food_items, lowBound = 0, cat='Continous')

### objective function is the cost which we are trying to minimize

In [7]:
#objective function si the cost which we are trying to minimize
prob += lpSum([costs[i]*food_vars[i] for i in food_items])

### Declare constraints each bound by minimum and maximum threshold provided

In [8]:
#define calorie constraints
prob += lpSum([calories[f] * food_vars[f] for f in food_items]) >= 1500, "CalorieMinimum"
prob += lpSum([calories[f] * food_vars[f] for f in food_items]) <= 2500, "CalorieMaximum"

In [9]:
# Cholesterol
prob += lpSum([cholesterol[f] * food_vars[f] for f in food_items]) >= 30.0
prob += lpSum([cholesterol[f] * food_vars[f] for f in food_items]) <= 240.0

In [10]:
# Fat
prob += lpSum([fat[f] * food_vars[f] for f in food_items]) >= 20.0
prob += lpSum([fat[f] * food_vars[f] for f in food_items]) <= 70.0

# Sodium
prob += lpSum([sodium[f] * food_vars[f] for f in food_items]) >= 800.0
prob += lpSum([sodium[f] * food_vars[f] for f in food_items]) <= 2000.0

# Carbs
prob += lpSum([carbs[f] * food_vars[f] for f in food_items]) >= 130.0
prob += lpSum([carbs[f] * food_vars[f] for f in food_items]) <= 450.0

# Fiber
prob += lpSum([fiber[f] * food_vars[f] for f in food_items]) >= 125.0
prob += lpSum([fiber[f] * food_vars[f] for f in food_items]) <= 250.0

# Protein
prob += lpSum([protein[f] * food_vars[f] for f in food_items]) >= 60.0
prob += lpSum([protein[f] * food_vars[f] for f in food_items]) <= 100.0

# Vitamin A
prob += lpSum([vit_A[f] * food_vars[f] for f in food_items]) >= 1000.0
prob += lpSum([vit_A[f] * food_vars[f] for f in food_items]) <= 10000.0

# Vitamin C
prob += lpSum([vit_C[f] * food_vars[f] for f in food_items]) >= 400.0
prob += lpSum([vit_C[f] * food_vars[f] for f in food_items]) <= 5000.0

# Calcium
prob += lpSum([calcium[f] * food_vars[f] for f in food_items]) >= 700.0
prob += lpSum([calcium[f] * food_vars[f] for f in food_items]) <= 1500.0

# Iron
prob += lpSum([iron[f] * food_vars[f] for f in food_items]) >= 10.0
prob += lpSum([iron[f] * food_vars[f] for f in food_items]) <= 40.0

### solve linear prob 

In [11]:
prob.solve()

1

#### there are 5 status codes:

####    Not Solved: Status prior to solving the problem.
####    Optimal: An optimal solution has been found.
####    Infeasible: There are no feasible solutions (e.g. if you set the constraints x <= 1 and x >=2).
####    Unbounded: The constraints are not bounded, maximising the solution will tend towards infinity (e.g. ####    if the only constraint was x >= 3).
####    Undefined: The optimal solution may exist but may not have been found.


In [12]:
LpStatus[prob.status]

'Optimal'

## Derived optimal solution with mimium cost

### Food items which formed part of optimal solution at minimum cost

In [13]:

prob.variables()


[food_2%_Lowfat_Milk,
 food_3.3%_Fat,Whole_Milk,
 food_Apple,Raw,W_Skin,
 food_Apple_Pie,
 food_Bagels,
 food_Banana,
 food_Beanbacn_Soup,W_Watr,
 food_Bologna,Turkey,
 food_Butter,Regular,
 food_Cap'N_Crunch,
 food_Carrots,Raw,
 food_Celery,_Raw,
 food_Cheddar_Cheese,
 food_Cheerios,
 food_Chicknoodl_Soup,
 food_Chocolate_Chip_Cookies,
 food_Corn_Flks,_Kellogg'S,
 food_Couscous,
 food_Crm_Mshrm_Soup,W_Mlk,
 food_Frankfurter,_Beef,
 food_Frozen_Broccoli,
 food_Frozen_Corn,
 food_Grapes,
 food_Ham,Sliced,Extralean,
 food_Hamburger_W_Toppings,
 food_Hotdog,_Plain,
 food_Kielbasa,Prk,
 food_Kiwifruit,Raw,Fresh,
 food_Lettuce,Iceberg,Raw,
 food_Macaroni,Ckd,
 food_Malt_O_Meal,Choc,
 food_New_E_Clamchwd,W_Mlk,
 food_Neweng_Clamchwd,
 food_Oatmeal,
 food_Oatmeal_Cookies,
 food_Oranges,
 food_Peanut_Butter,
 food_Peppers,_Sweet,_Raw,
 food_Pizza_W_Pepperoni,
 food_Poached_Eggs,
 food_Popcorn,Air_Popped,
 food_Pork,
 food_Potato_Chips,Bbqflvr,
 food_Potatoes,_Baked,
 food_Pretzels,
 food_Raisi

### food item with non zero coefficients - air-popped popcorn, poached eggs, oranges, raw iceberg lettuce, raw celery, and frozen broccoli.

In [14]:
for i in prob.variables():
    if i.varValue>0:
        print(i.name, "=", i.varValue)

food_Celery,_Raw = 52.64371
food_Frozen_Broccoli = 0.25960653
food_Lettuce,Iceberg,Raw = 63.988506
food_Oranges = 2.2929389
food_Poached_Eggs = 0.14184397
food_Popcorn,Air_Popped = 13.869322


### As per the solver, optimal diet includes 52.6 servings of celery, 0.25 of forzenbroccoli etc...

In [15]:
print("The total cost of balanced diet is: ${}".format(round(value(prob.objective),2)))

The total cost of balanced diet is: $4.34


In [16]:
prob.objective

0.23*food_2%_Lowfat_Milk + 0.16*food_3.3%_Fat,Whole_Milk + 0.24*food_Apple,Raw,W_Skin + 0.16*food_Apple_Pie + 0.16*food_Bagels + 0.15*food_Banana + 0.67*food_Beanbacn_Soup,W_Watr + 0.15*food_Bologna,Turkey + 0.05*food_Butter,Regular + 0.31*food_Cap'N_Crunch + 0.07*food_Carrots,Raw + 0.04*food_Celery,_Raw + 0.25*food_Cheddar_Cheese + 0.28*food_Cheerios + 0.39*food_Chicknoodl_Soup + 0.03*food_Chocolate_Chip_Cookies + 0.28*food_Corn_Flks,_Kellogg'S + 0.39*food_Couscous + 0.65*food_Crm_Mshrm_Soup,W_Mlk + 0.27*food_Frankfurter,_Beef + 0.16*food_Frozen_Broccoli + 0.18*food_Frozen_Corn + 0.32*food_Grapes + 0.33*food_Ham,Sliced,Extralean + 0.83*food_Hamburger_W_Toppings + 0.31*food_Hotdog,_Plain + 0.15*food_Kielbasa,Prk + 0.49*food_Kiwifruit,Raw,Fresh + 0.02*food_Lettuce,Iceberg,Raw + 0.17*food_Macaroni,Ckd + 0.52*food_Malt_O_Meal,Choc + 0.99*food_New_E_Clamchwd,W_Mlk + 0.75*food_Neweng_Clamchwd + 0.82*food_Oatmeal + 0.09*food_Oatmeal_Cookies + 0.15*food_Oranges + 0.07*food_Peanut_Butter + 0.5